src/backend/optimizer/path/allpaths.c

```diff
@@ -4434,6 +4444,165 @@ is_query_contain_limit_groupby(Query *parse)
 	return false;
 }
 
+typedef struct check_dml_context
+{
+	PlannerInfo *cteroot;
+	Index		m_query_level;	/* absolute query level */
+}			check_dml_context;
+
+/*
+ * Returns true if any subquery of given Query has any non-SELECT commandType.
+ */
+static bool
+query_contains_dml_walker(Node *node, check_dml_context * ctx)
+{
+	if (node == NULL)
+		return false;
+
+	if (IsA(node, Query))
+	{
+		Query	   *query = (Query *) node;
+		ListCell   *lc;
+
+		if (query->commandType != CMD_SELECT)
+			return true;
+
+		foreach(lc, query->rtable)
+		{
+			RangeTblEntry *rte = (RangeTblEntry *) lfirst(lc);
+
+			if (rte->rtekind == RTE_SUBQUERY)
+			{
+				ctx->m_query_level++;
+				if (query_contains_dml_walker((Node *) rte->subquery, ctx))
+					return true;
+				ctx->m_query_level--;
+			}
+			else if (rte->rtekind == RTE_CTE)
+			{
+				/*
+				 * rte->ctelevelsup is equal to number of levels up from the
+				 * level where it's referenced to the level where its CTE is
+				 * defined. Difference ctx->m_query_level - rte->ctelevelsup
+				 * stands for absolute query level, where the CTE is defined.
+				 * In case if this level is higher than absolute level where
+				 * the original CTE (last CTE and its root, which the walker
+				 * was last called with) is defined, we are able to derive the
+				 * root of CTE related to current rte. Based on its cteList it
+				 * is possible to investigate the CTE. Otherwise, current CTE
+				 * is defined at levels lower than original is. In this case
+				 * the subroot with this CTE do not exist and we will get to
+				 * this CTE by falling back to query_tree_walker.
+				 * E.x. WITH cte1 AS (INSERT into...) SELECT * FROM
+				 * (WITH cte2 as (SELECT * from cte1) SELECT * FROM
+				 * (WITH cte3 as (SELECT * FROM cte2) SELECT * FROM
+				 * cte3 a join cte3 b using (i)) foo) bar;
+				 * Here cte3 is an entry point for the walker with
+				 * m_query_level = 4. cte3 references the cte2, whose
+				 * ctelevelsup is 2. It means that cte2 is defined at 2 level
+				 * up from the current level. This level is higher than level
+				 * where cte3 is defined. cte3 is defined at level 3 and cte2
+				 * is at level 2. In this case we can derive parent root,
+				 * whose cteList contains query of cte2.
+				 */
+				if (rte->ctelevelsup < ctx->m_query_level &&
+					(ctx->m_query_level - rte->ctelevelsup) <= ctx->cteroot->query_level)
+				{
+					ListCell   *lcn;
+					PlannerInfo *cteroot = ctx->cteroot;
+					Index		query_level = ctx->m_query_level;
+					Index		levelsup = ctx->cteroot->query_level -
+					(ctx->m_query_level - rte->ctelevelsup);
```

Comment:
Maybe replace member access with variables defined earlier?
```suggestion
					Index		query_level = ctx->m_query_level;
					Index		levelsup = cteroot->query_level -
					(query_level - rte->ctelevelsup);
```